test/model/autocomplete_test.dart

```diff
@@ -178,8 +184,106 @@ void main() {
-    doTest('If @_chris is around, please ask him.^', null); // @ sign is too far away from cursor
+
+    // "#" sign can be (3 + 2 * maxChannelName) utf-16 code units
+    // away to the left of cursor.
+    doTest('If ~@chris^ is around, please ask him.', mention('chris'), maxChannelName: 10);
+    doTest('If ~@_chris is^ around, please ask him.', silentMention('chris is'), maxChannelName: 10);
+    doTest('If @chris is around, please ask him.^', null, maxChannelName: 10);
+    doTest('If @_chris is around, please ask him.^', null, maxChannelName: 10);
+
+    // #channel link.
+
+    doTest('^#', null);
+    doTest('^#abc', null);
+    doTest('#abc', null); // (no cursor)
+
+    doTest('~#^', channelLink(''));
+    doTest('~##^', channelLink('#'));
+    doTest('~#abc^', channelLink('abc'));
+    doTest('~#abc ^', channelLink('abc '));
+    doTest('~#abc def^', channelLink('abc def'));
+
+    // Accept space before channel link syntax.
+    doTest(' ~#abc^', channelLink('abc'));
+    doTest('xyz ~#abc^', channelLink('abc'));
+
+    // Accept punctuations before channel link syntax.
+    doTest(':~#abc^', channelLink('abc'));
+    doTest('!~#abc^', channelLink('abc'));
+    doTest(',~#abc^', channelLink('abc'));
+    doTest('.~#abc^', channelLink('abc'));
+    doTest('(~#abc^', channelLink('abc')); doTest(')~#abc^', channelLink('abc'));
+    doTest('{~#abc^', channelLink('abc')); doTest('}~#abc^', channelLink('abc'));
+    doTest('[~#abc^', channelLink('abc')); doTest(']~#abc^', channelLink('abc'));
+    doTest('“~#abc^', channelLink('abc')); doTest('”~#abc^', channelLink('abc'));
+    doTest('«~#abc^', channelLink('abc')); doTest('»~#abc^', channelLink('abc'));
+    // … and other punctuations except '#' and '@':
+    doTest('~##abc^', channelLink('#abc'));
+    doTest('~@#abc^', mention('#abc'));
+
+    // Avoid other characters before channel link syntax.
+    doTest('+#abc^', null);
+    doTest('=#abc^', null);
+    doTest('\$#abc^', null);
+    doTest('zulip/zulip-flutter#124^', null);
+    doTest('XYZ#abc^', null);
+    doTest('xyz#abc^', null);
+    // … but
+    doTest('~#xyz#abc^', channelLink('xyz#abc'));
+
+    // Avoid leading space character in query.
+    doTest('# ^', null);
+    doTest('# abc^', null);
+
+    // Avoid line-break characters in query.
+    doTest('#\n^', null); doTest('#a\n^', null); doTest('#\na^', null); doTest('#a\nb^', null);
+    doTest('#\r^', null); doTest('#a\r^', null); doTest('#\ra^', null); doTest('#a\rb^', null);
+    doTest('#\r\n^', null); doTest('#a\r\n^', null); doTest('#\r\na^', null); doTest('#a\r\nb^', null);
+
+    // Allow all other sorts of characters in query.
+    doTest('~#\u0000^', channelLink('\u0000')); // control
+    doTest('~#\u061C^', channelLink('\u061C')); // format character
+    doTest('~#\u0600^', channelLink('\u0600')); // format
+    doTest('~#\uD834^', channelLink('\uD834')); // leading surrogate
```

Comment:
Yuck. Are these really desirable? 🙂 

We don't need to add logic to exclude these cases; I don't think they're important. But I'd like these tests to focus on the behavior we actually *want*, not on the behavior that our implementation happens to have by accident. That way we're not spending attention on thinking about these cases either in the implementation or the tests.